### PR TITLE
Simplify port API and add a data source task

### PR DESCRIPTION
**Pull request by jbeezley**
Besides additional testing, this PR changes output ports so that they only output a single data type.  Input ports can still accept multiple types.  Two ports are compatible if the output type is (or a subclass of) one of the possible input types.

I also added a method that makes it easier to execute tasks from the commandline by generating one off
source tasks.
```python
source = Task.create_source(data)
plotter = GeopandasPlot()
plotter.set_input(port=source.get_output())
plotter.run()
```

### CODE REVIEW

**Comment by jbeezley** [comment]
The curse of random bad VM's on travis returns.  Waiting on https://github.com/travis-ci/travis-ci/issues/3561 to move to containers.

**Comment by mgrauer** [comment]
@jbeezley I'm trying to provide something approaching QA here.  

Can you tell me the reason for output ports being changed to only have a single output?  

And why did you remove the explicit use of data registry?  For looser coupling or something else?

Your tests look good and comprehensive.  I think I'd be able to give better QA if I understood some of the use cases or driving requirements.

File: gaia/core/task.py [hunk 1]
```diff
@@ -132,9 +129,58 @@ def _reset(self, *args):
...
+        """Set dirty state on all downstream tasks."""
+        if isdirty:
+            for name in self.outputs:
+                task = self.get_output_task(name=name)
+                if task:
+                    task.dirty = True
+
+    @classmethod
+    def create_source(cls, data):
+        """Generate a task that supplies the given data as output.
```
Comment (aashish24) [inline gaia/core/task.py]: +1

File: gaia/pandas/geopandas_data.py [hunk 2]
```diff
@@ -2,10 +2,10 @@
 
 from geopandas import GeoDataFrame as _DataFrame
 
-from gaia.core.data import registry as data_registry
+from gaia.core.data import GeospatialData
 
 
-class GeopandasDataFrame(data_registry['GeospatialData'], _DataFrame):
+class GeopandasDataFrame(GeospatialData, _DataFrame):
```
Comment (aashish24) [inline gaia/pandas/geopandas_data.py]: Can you explain this change?

**Comment by jbeezley** [comment]
I just thought of one additional change I want to make here to simplify the data API.  I'll push it later today and attempt to explain the motivation behind the changes.

**Comment by jbeezley** [comment]
Also, the travis containers are randomly unable to install fiona.  :angry:

**Comment by aashish24** [comment]
@mgrauer what @jbeezley changed is the fact that one outport can only have one data type.

File: gaia/core/port.py [hunk 3]
```diff
@@ -83,24 +81,31 @@ class InputPort(Port):
...
     def connect(self, other):
-        """Negotiate a common format and connect two tasks together.
+        """Check for a common format and connect two tasks together.
 
         Calls connect on the output port.
         """
         if not isinstance(other, OutputPort):
             raise TypeError("Invalid connection with {0}".format(other))
-        return super(InputPort, self).connect(other)
+        other.connect(self)
```
Comment (aashish24) [inline gaia/core/port.py]: how does it work?

Comment (aashish24) [inline gaia/core/port.py]: nvm

**Comment by aashish24** [comment]
I haven't ran the tests yet.

**Comment by jbeezley** [comment]
Surry that I never had a chance to explain this.  We should meet up some time with @mgrauer to discuss the objectives going forward once I get back; especially considering the changes coming that will tie in Romanesco.

**Pull request merged**